parsec/backend/postgresql/pki.py

```diff
@@ -0,0 +1,439 @@
+            for row in rep:
+                enrollment_state = row["enrollment_state"]
+                if enrollment_state == PkiEnrollmentStatus.SUBMITTED.value:
+                    if force:
+                        await conn.execute(
+                            *_q_cancel_pki_enrollment(
+                                organization_id=organization_id.str,
+                                enrollment_id=row["enrollment_id"],
+                                enrollment_state=PkiEnrollmentStatus.CANCELLED.value,
+                                cancelled_on=submitted_on,
+                            )
+                        )
+                        await send_signal(
+                            conn,
+                            BackendEvent.PKI_ENROLLMENT_UPDATED,
+                            organization_id=organization_id,
+                        )
+
+                    else:
+                        raise PkiEnrollmentCertificateAlreadySubmittedError(
+                            submitted_on=row["submitted_on"]
+                        )
+                elif enrollment_state in [
+                    PkiEnrollmentStatus.REJECTED.value,
+                    PkiEnrollmentStatus.CANCELLED.value,
+                ]:
+                    # Previous attempt was unsuccessful, so we are clear to submit a new attempt !
+                    pass
+                elif enrollment_state == PkiEnrollmentStatus.ACCEPTED.value:
+                    # Previous attempt end successfully, we are not allowed to submit
+                    # unless the created user has been revoked
+                    assert row["accepted"] is not None and row["accepter"] is not None
+
+                    row = await conn.fetchrow(
+                        *_q_get_user_from_device_id(
+                            organization_id=organization_id.str,
+                            device_id=row["accepted"],
+                            # now=submitted_on,
+                        )
+                    )
+                    user = User(
+                        user_id=UserID(row["user_id"]),
+                        human_handle=None,
+                        profile=UserProfile(row["profile"]),
+                        user_certificate=row["user_certificate"],
+                        redacted_user_certificate=row["redacted_user_certificate"],
+                        user_certifier=None,
+                        created_on=row["created_on"],
+                        revoked_on=row["revoked_on"],
+                        revoked_user_certificate=row["revoked_user_certificate"],
+                        revoked_user_certifier=None,
+                    )
+                    # if row and row.get("revoked_on") and row.get("revoked_on") < submitted_on:
+                    if not user.is_revoked():
+                        raise PkiEnrollmentAlreadyEnrolledError(submitted_on)
+                else:
+                    assert False
+            await conn.execute(
```

Comment:
should check the return value and exception to handle the (unlikely !) case of two concurrent submits with the same `enrollement_id`, hence one succeed and the other raising an `UniqueViolationError` (see what is done in `postgresql/user_queries/create.py` when creating user/device)